### ISSUE TITLE
SSH Passphrase input fails strangely in rhel 6.5-6.6

### ISSUE THREAD

**Issue by sona1111** [pexpect/pexpect#192]
I have been trying to resolve this issue through various communications (ex. http://stackoverflow.com/questions/28823639/how-to-automatically-input-ssh-private-key-passphrase-with-pexpect) , but I have not had success for a significant time duration so I am trying to ask at the source now. As part of a program I need to automatically enter a passphrase for an SSH key. I built a simple pexpect script such as, which should mount via sshfs:

    proc = pexpect.spawn('sshfs %s@%s:%s -o identityfile=...') #many options, unrelated
    proc.expect([pexpect.EOF, 'Enter passphrase for key.*', pexpect.TIMEOUT], timeout=30)
    if proc.match_index == 1:
        proc.sendline('thepassphrase')

On ubuntu server distributions, this works correctly, but when switching to rhel 6.5 or 6.6 (ones I have tested so far) I get the issue. What happens is that the 'enter passphrase for key' is caught and the line is seemingly sent, but then the process exits before the mounting is complete. I get the 'transport endpoint is not connected' error - an error that would happen if I tried running the command but then sent a SIGINT before entering the passphrase. I have tried various combinations of waits and communicates but after being on it for literally days I think it may just be a bug. 

(note that I do mean _passphrase_ and not _password_, I _am_ using a SSH private key, it just has a passphrase)

Let me know if you need more information.

**Comment by jquast** [comment]
Try to ``expect`` all the way up to the final character before sending input.

```
- 'Enter passphrase for key.*'
+ 'Enter passphrase for key.*: '
```

What version of python are you using on redhat? If I recall, redhat still ships with python2.6 whose final release is about 7 years old.

Anyway I could only reproduce this issue, seemingly, by creating an account with redhat.com to download a 30-day evaluation copy and installing fuse/sshfs and somehow making up the same environment you've got over there.  If you sent me a virtualbox image with a ./test.py in a home folder that reproduced the issue exactly you would save me a whole lot of trouble.

**Comment by sona1111** [comment]
jquast, I really appreciate the prompt reply. 

I will attempt the solution you have provided first. 

I have built python 2.7 on rhel and this is the version I have used for all testing. The machines I use are AMIs on AWS but I can see if it is possible to obtain a local version to test with, do not bother trying to set this up yourself. ;)

Thank you.

**Comment by sona1111** [comment]
jquast, 

Unfortunately I have tried the modified expect string without any difference. I have sent you an email with a link to the VDI. If you are ever feeling bored you may take a look at it. 

Thank you.

**Comment by jquast** [comment]
I got the VDI, thank you, I was able to reproduce.

I don't understand the cause, but I thought I'd reply with a workaround anyway since you seem to be struggling -- with the given example script, ``sshfs`` process does not remain after exit.

A simple enough workaround is to spawn an *interactive* bash shell, send the command sequence and exit nicely:

**sshfs process remains after exit**

```

passphrase = 'some passphrase'
cmd = "sudo -H -u somebody sshfs somebody@somewhere:/somewhere-else /home/somebody/testmount -o StrictHostKeychecking=no -o nonempty -o reconnect -o workaround=all -o IdentityFile=/somekey -o follow_symlinks -o cache=no"

bash = pexpect.spawn('bash', echo=False)
bash.sendline('echo READY')
bash.expect_exact('READY')
bash.sendline(cmd)
bash.expect_exact('Enter passphrase for key')
bash.sendline(passphrase)
bash.sendline('echo COMPLETE')
bash.expect_exact('COMPLETE')
bash.sendline('exit')
bash.expect_exact(pexpect.EOF)
```

**sshfs process does not remain after exit**

```
bash2 = pexpect.spawn('bash', ['-c', cmd])
bash2.expect_exact('Enter passphrase')
bash2.sendline(passphrase)
```

I have a feeling this has something to do with signals, such as SIGCHLD, but I don't know for sure.

**Comment by sona1111** [comment]
jquast, 

I am sorry to take so long to respond over these few days. First of all, I really, really appreciate the workaround. I spent a very long time on this and this type of solution is definitely going into my collection of useful snippets. Because no one else seems to care or even notice the issue, and your workaround is ok for me, feel free to keep it as a low priority issue. Also, if you would like to respond to the stackoverflow question that I linked earlier I will mark your answer as accepted. 

Thank you.
